### PR TITLE
Tests and fix for infinite loop when setting shape to nil in MGLMapViewDelegate methods

### PR DESCRIPTION
**Pull request by julianrex**
Addresses #11180, #5833.
Adds integration tests and work-around to stop crash.

### CODE REVIEW

File: src/mbgl/map/transform.cpp [hunk 1]
```diff
@@ -595,12 +595,16 @@ void Transform::startTransition(const CameraOptions& camera,
             }
             observer.onCameraIsChanging();
         } else {
-            transitionFinishFn();
+            auto finish = transitionFinishFn;
```
Comment (jfirebaugh) [inline src/mbgl/map/transform.cpp]: Can you add comments explaining why this code is structured this way?

File: src/mbgl/map/transform.cpp [hunk 2]
```diff
@@ -595,12 +595,16 @@ void Transform::startTransition(const CameraOptions& camera,
...
+
             transitionFinishFn = nullptr;
 
             // This callback gets destroyed here,
             // we can only return after this point.
```
Comment (jfirebaugh) [inline src/mbgl/map/transform.cpp]: I think this comment was meant to serve as a reminder not to change this lambda such that any code executes "after this point". But this change does that. Is it safe? If so, please update this comment to explain why.

**Comment by julianrex** [comment]
Bisecting and testing with @jmkiley's test case, it looks like this issue was introduced somewhere in:

https://github.com/mapbox/mapbox-gl-native/compare/39a732d7ae3cb1b927d94c4b1154b42d9565356a...b6d56ad634e2b3048e97bedd9f674aa4ec975453

/cc @ivovandongen

**Comment by ivovandongen** [comment]
@julianrex Looking at the test case I wonder if this is an actual bug. If the change of a source is to trigger an update of the transitions and thus callback with `onCameraDidChange()`, this would be user error. Preventable by checking if the action has already been performed or clearing the callback when executing. 

If the update of a source shouldn't trigger `onCameraDidChange()` in any case, we should address that problem at the root.

**Comment by julianrex** [comment]
@jfirebaugh @ivovandongen What functionality should be disallowed in the various callbacks (e.g. onCameraDidChange)? 

It sounds like solution should be a two part fix: 
1. The proposed solution to stop the infinite loop if it gets to that point, and 
2. raising an exception in client (iOS, Android) code for those operations that we should disallow?

What are you thoughts on:  1) could make it into the `boba` release, and 2) is tackled in a patch release.

/cc @tobrun as this may affect Android too.

**Comment by julianrex** [comment]
Side-note: I've extracted the test changes into a new PR: https://github.com/mapbox/mapbox-gl-native/pull/11649, and update this PR when that has been merged.

**Comment by julianrex** [comment]
Pushed changes that I believe addresses #5833.

**Comment by julianrex** [comment]
Rebased.

File: src/mbgl/map/transform.cpp [hunk 3]
```diff
@@ -624,8 +628,35 @@ bool Transform::inTransition() const {
...
+        // the finish lambda (going via a temporary as above)
+        auto finish = std::move(transitionFinishFn);
+
+        transitionFinishFn = nullptr;
+        transitionFrameFn = nullptr;
+
+        if (finish) {
+            finish();
+        }
+    }
```
Comment (jfirebaugh) [inline src/mbgl/map/transform.cpp]: Nit: `}` and `else if ...` on the same line.

File: src/mbgl/map/transform.cpp [hunk 4]
```diff
@@ -624,8 +628,35 @@ bool Transform::inTransition() const {
...
+    else if (!transitionFrameFn) {
```
[the remark this answers is not in the document]
Comment (jfirebaugh) [reply src/mbgl/map/transform.cpp]: Probably not -- I can imagine a situation where the camera change notification triggers the start of another transition, in which case `transitionFrameFn` would be legitimately non-null.

**Comment by julianrex** [comment]
Rebasing due to release-boba merge.

**Comment by friedbunny** [comment]
@julianrex I think this needs to be retargeted at `master`?

**Pull request merged**